### ISSUE TITLE
Make dependencies conditional for optional features

### ISSUE THREAD

**Issue by natefoo** [galaxyproject/galaxy#7320]
Galaxy has a feature for automatically adding to its requirements depending on features that are enabled. Many new dependencies have been added to the fixed requirements that should probably be conditional requirements. Doing so will drastically decrease the size of e.g. Docker images where these features are not enabled by default, as well as decrease initial startup times.

As far as I can tell, these requirements from the current `dev` are all for optional cloud integration features and should be made conditional. Doing so would decrease the size of the virtualenv by (in my testing) 183MB:

```
azure-common==1.1.14
azure-cosmosdb-nspkg==2.0.2
azure-cosmosdb-table==1.0.4
azure-mgmt-compute==4.0.1
azure-mgmt-devtestlabs==2.2.0
azure-mgmt-network==2.1.0
azure-mgmt-nspkg==3.0.2
azure-mgmt-resource==2.0.0
azure-mgmt-storage==2.0.0
azure-nspkg==3.0.2
azure-storage-blob==1.3.1
azure-storage-common==1.4.0
azure-storage-nspkg==3.1.0
boto3==1.7.84
boto==2.49.0
botocore==1.10.84
cloudauthz==0.2.0
cloudbridge==1.0.2
keystoneauth1==3.11.2
msrest==0.5.5
msrestazure==0.5.0
netaddr==0.7.19
netifaces==0.10.9
oauthlib==3.0.1
openstacksdk==0.17.0
os-client-config==1.31.2
os-service-types==1.5.0
osc-lib==1.12.0
oslo.config==6.8.0
oslo.context==2.22.0
oslo.i18n==3.23.0
oslo.log==3.42.2
oslo.serialization==2.28.1
oslo.utils==3.40.1
python-cinderclient==4.0.0
python-glanceclient==2.12.0
python-keystoneclient==3.17.0
python-neutronclient==6.9.0
python-novaclient==11.0.0
python-openid==2.2.5
python-swiftclient==3.6.0
requests-oauthlib==1.2.0
s3transfer==0.1.13
social-auth-core[openidconnect]==1.5.0
```

Ping @afgane @VJalili - any others?

**Comment by VJalili** [comment]
`social-auth-core[openidconnect]==1.5.0` is required if OIDC is enabled:

https://github.com/galaxyproject/galaxy/blob/0df943f5e3d0229d5d355e842b15cf6cdc662914/config/galaxy.yml.sample#L1429-L1430

Azure, CloudBridge, and etc. are required for ObjectStore and the [Cloud API](https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/webapps/galaxy/api/cloud.py): 
- Cloud may not be set on ObjectStore, so you would not need to install these dependencies if Cloud is not enabled; 
- however, I am not sure how/if we can enable/disable APIs?

**Comment by natefoo** [comment]
@VJalili that's the point, see https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/dependencies/__init__.py

From here you can check if e.g. `enable_oidc` is set, and indicate whether the dependency is needed accordingly.

[Here's a bit checking whether Azure Blobs are enabled in the object store](https://github.com/galaxyproject/galaxy/blob/0df943f5e3d0229d5d355e842b15cf6cdc662914/lib/galaxy/dependencies/__init__.py#L130), presumably this would still work, and the azure object store requirements just need to be moved from `pinned-requirements.txt` to `conditional-requirements.txt`.

You also have to [catch ImportErrors](https://github.com/galaxyproject/galaxy/blob/0df943f5e3d0229d5d355e842b15cf6cdc662914/lib/galaxy/objectstore/azure_blob.py#L12) so that tests won't fail.

**Comment by VJalili** [comment]
I am assuming it is not possible to enable/disable APIs in general; if that is correct, then since Azure is required for the Cloud API, it should always install.

**Comment by VJalili** [comment]
However, not all Azure features are required by the `Cloud` api (e.g., `azure-mgmt-compute`), so not all of those dependencies are required. Therefore, probably we can find a workaround (in cloudbridge maybe), and "require" only the dependencies essential for a subset of cloudbridge functionality. For instance, if I am using only "storage" functionalities from cloudbirdge, I would only have the strictly required dependencies installed.

**Comment by nsoranzo** [comment]
@VJalili You can specify optional dependencies of a Python packages using extras: https://setuptools.readthedocs.io/en/latest/setuptools.html#declaring-extras-optional-features-with-their-own-dependencies

**Comment by natefoo** [comment]
It should be possible to disable the Cloud API by catching import failures the same way as is done for the azure blob object store. That said, if it's the case that the Cloud API should always be enabled, we'd need to work around that - maybe a configuration option for which clouds are enabled?

**Comment by VJalili** [comment]
@natefoo So I guess we have two questions: 

1- do we want to be able to enable/disable APIs? (so different Galaxy instances having different set of APIs enabled);
2- do we want to have the payloads of APIs configurable? (e.g., one instance of Galaxy accepts AWS for its Cloud API, another instance accepts AWS and Azure). 

@nsoranzo Do you think _extras_ should be first defined in Cloudbridge, and then leveraged in Galaxy?

**Comment by nsoranzo** [comment]
@VJalili Yes, that's how you can have a core `cloudbridge` package and optional "features" available with the extra packages. As an example, try to `pip install requests` and `requests[security]` in a virtualenv.

**Comment by VJalili** [comment]
@nsoranzo Thanks for the clarification; that is what I guessed. Currently in cloudbridge, I don't think we have "feature-specific" list of requirements (e.g., I want to use only Azure provider, and for data storage only, what are the minimal list of requirements?): 

https://github.com/CloudVE/cloudbridge/blob/75314f23e0d0e2cb012227c37b4dcfb10c715200/setup.py#L74-L78

**Comment by nuwang** [comment]
@nsoranzo The problem that we ran into a while back with extras is that it only allows you to add extra packages over the base requirements, not to remove or redefine what the base requirements are. Therefore, if we make all cloud providers extras, then a simple `pip install cloudbridge` would result in something that is pretty much non-functional, since all providers would be extras. It seems like what we need is a way to specify a default set of extras, or to remove packages from the default install, not just add them. However, there still doesn't appear to be any support for this: https://github.com/pypa/setuptools/issues/1139
Do you know of any other workarounds?

**Comment by nsoranzo** [comment]
The only other way I know of is creating finer-grade subpackages (e.g. `cloudbridge-core`, `cloudbridge-azure`, `cloudbridge-aws`,...) and an "empty" `cloudbridge` requiring all of them. Then Galaxy could conditionally depend on the subpackages.
But this requires more work to create the various subpackages and keep them in sync.

**Comment by natefoo** [comment]
@VJalili I'd say yes to both of these questions.

**Comment by VJalili** [comment]
@natefoo sounds good. Do we have such features currently that we can leverage to implement these? (i.e., can a Galaxy admin currently enable/disable some APIs, and probably tweak what is an acceptable payload for an API?)

**Comment by martenson** [comment]
@VJalili something like this is a start: https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/webapps/galaxy/api/users.py#L185

**Comment by VJalili** [comment]
@martenson sounds good. Thank you!
